include/nil/blueprint/components/algebra/fields/plonk/logic_or_flag.hpp

```diff
@@ -0,0 +1,233 @@
+                        bp.add_gate(first_selector_index, {constraint_1, constraint_2, constraint_3, constraint_4, constraint_5, constraint_6, constraint_7});
+                    }
+
+                }
+
+                template<typename BlueprintFieldType, typename ArithmetizationParams, std::uint32_t WitnessesAmount>
+                void generate_copy_constraints(
+                    const plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>& component,
+                    circuit<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &bp,
+                    assignment<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &assignment, 
+                    const typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::input_type &instance_input, 
+                    const std::uint32_t start_row_index){
+
+                    std::size_t row = start_row_index;
+                    using var = typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::var;
+
+                    bp.add_copy_constraint({var(component.W(0), row, false), instance_input.x});
+                    bp.add_copy_constraint({var(component.W(1), row, false), instance_input.y});
+                }
+
+                template<typename BlueprintFieldType, typename ArithmetizationParams, std::uint32_t WitnessesAmount>
+                typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::result_type generate_circuit(
+                    const plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>& component,
+                    circuit<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &bp,
+                    assignment<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &assignment, 
+                    const typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::input_type &instance_input, 
+                    const std::uint32_t start_row_index){
+
+                    auto selector_iterator = assignment.find_selector(component);
+                    std::size_t first_selector_index;
+
+                    if (selector_iterator == assignment.selectors_end()) {
+                        first_selector_index = assignment.allocate_selector(component, component.gates_amount);
+                        generate_gates(component, bp, assignment, instance_input, first_selector_index);
+                    } else {
+                        first_selector_index = selector_iterator->second;
+                    }
+                    assignment.enable_selector(first_selector_index, start_row_index + component.rows_amount == 3 ? 1 : 0);
```

Comment:
This is wrong: 
`start_row_index + component.rows_amount == 3 ? 1 : 0`
Is evaluated as `((start_row_index + component.rows_amount) == 3) ? 1 : 0`.
Note that `==` has lower precedence than `+`, and `?` has one of the lowest levels of operator precedence in C++ (only `,` has lower precedence).
You get (un?)lucky in your tests because `start_row_index` is always zero.
I should probably think about randomizing that for tests.

Reply:
[my explanation is probably redundant, as you got this right in the other component, sorry]